File: java_challenges/main/Matrix.java
import java.util.Arrays;

public class Matrix {

    Boolean SolveRecursively = true;

    int[][] maze = new int[][]{{}};
    int rows = 0;
    int columns = 0;

    public Matrix(int[][] maze) {
        this.maze = maze;
        rows = maze.length;
        columns = maze[0].length;
    }

    public int solve() {
        int paths;
        showStart();
        paths = SolveRecursively ? countPathsRecursive(maze, 0, 0) : countPaths();
        showResult(paths);
        return paths;
    }

    private int countPaths() {
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < columns; j++) {
                if (i == 0 && j > 0 && maze[i][j] == 1) maze[0][j] = maze[0][j - 1];
                else if (i > 0 && j == 0 && maze[i][j] == 1) maze[i][0] = maze[i - 1][0];
                else if (i > 0 && j > 0 && maze[i][j] == 1) maze[i][j] = maze[i - 1][j] + maze[i][j - 1];
            }
        }
        return maze[rows - 1][columns - 1];
    }

    private int countPathsRecursive(int[][] maze, int i, int j) {
        if (i >= rows || j >= columns) {
            return 0;
        } else if (maze[i][j] == 0) {
            return 0;
        } else if (i == rows - 1 && j == columns - 1) {
            return 1;
        } else {
            return countPathsRecursive(maze, i + 1, j) + countPathsRecursive(maze, i, j + 1);
        }
    }

    private void showMaze() {
        for (int[] row : maze) {
            System.out.println("    " + Arrays.toString(row));
        }
    }

    private void showStart() {
        System.out.println("\nMaze: ");
        showMaze();
    }

    private void showResult(int paths) {
        if (!SolveRecursively) {
            System.out.println("Path count analysis:");
            showMaze();
        }
        System.out.println("result: " + paths);
    }
}

File: ruby_challenges/src/matrix.rb
require 'pp'

class Matrix

  def initialize(values)
    puts "\nCalculating paths for matrix:"
    @values = values
    show
  end

  def solve
    (0..(@values.length - 1)).each do |row|
      (0..(@values[row].length - 1)).each do |column|
        unless @values[row][column] == 0
          unless row == 0 and column == 0
            if row == 0
              @values[row][column] = @values[row][column - 1]
            elsif column == 0
              @values[row][column] = @values[row - 1][column]
            else
              @values[row][column] = @values[row - 1][column] + @values[row][column - 1]
            end
          end
        end
      end
    end
    puts 'Path analysis:'
    show
    result = @values[-1][-1]
    puts "Number of paths: #{result}"
    result
  end

  def show
    (0..(@values.length - 1)).each do |row|
      pp @values[row]
    end
  end

end